#4982 Fix behaviour for failed update of customer notes

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #4982
<!-- Id number of the GitHub issue this PR addresses. -->
Is related: #4984 

FluxC PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2149

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR fixes a bug of putting empty values for edited order, when customer notes update failed. The reason for that is that we used `Order#toDataModel` that was mapping only 3 values - rest of them we've been nulling.

This PRs also deprecates usage of `local id` or `OrderIdentifier` for Woo's `Order` class. Please see p91TBi-6dF-p2 for related discussion.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

#### TC1 Success case
Please smoke test updates of Order's `status` and `customer provided notes`. It should all correctly update values both locally (optimistically) and on backend.

#### TC2 Failure case
1. Checkout this branch
2. Change `fluxcversion` to `2148-35375ff88afa0bba8e7029f1c5e05bea13c2c275`. It contains changed logic that won't send event to backend but instead will emit failure after 5 seconds. [See FluxC temp PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2148/files#diff-ee210dfead61bbf73ac0446d71d96195a85a041bac7e76d1919c54359960bc41R61-R67).
3. Run the app, open any order
4. Change `Customer provided note`
5. Wait for 5 seconds on order details
6. **Assert that customer provided note reverted back to original value and that the rest of the order has meaningful values.**

### Images/gif

#### Before

https://user-images.githubusercontent.com/5845095/136818815-a2d19ce4-b08f-40d1-b5ff-174bef2ac169.mov

#### After

https://user-images.githubusercontent.com/5845095/136819300-265c7959-563c-478d-985f-002363695b1c.mov



<!-- Include before and after images or gifs when appropriate. -->


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Deprecate usage of local id and OrderIdentifier #4984

wzieba committed:
Update FluxC hash

wzieba committed:
Align to new update customer notes API

wzieba committed:
Align to new update order status API

wzieba committed:
Remove empty line

wzieba commented:
Hi @nbradbury , @ThomazFB ! This PR is ready for review alongside its FluxC part but please - let's don't merge it yet. I'd like to wait 1-2 days as it contains changes related to #4984 (discussed here p91TBi-6dF-p2) to give a change to everyone to contribute to discussion. Thanks!

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt:
```diff
@@ -138,11 +125,13 @@ class OrderDetailRepository @Inject constructor(
 
     suspend fun updateOrderStatus(
-        orderModel: WCOrderModel,
+        orderId: RemoteId,
```
I think we should rename this from "orderId" to "remoteOrderId" so it's clear what type of ID it is. Same goes for the other instances of "orderId".

wzieba replied:
True, but for this reason I've used type of `RemoteId` instead of long - I think it's easier this way to assert correct type. If we'll had `remoteOrderId: Long` it'd be easy to assign `local id` by mistake.

We could have `remoteOrderId: RemoteId` but I felt like the "remote" in the name might be redundant. WDYT?

nbradbury replied:
I think we should always be specific about it being a remote ID. We can keep the parameter name shorter by using `remoteId` instead of `remoteOrderId`.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt:
```diff
@@ -60,6 +53,9 @@ data class Order(
     val metaData: List<MetaData<String>>
 ) : Parcelable {
+    val id
+        get() = LocalOrRemoteId.RemoteId(this.remoteId)
```
As with the FluxC PR, I think we should be explicit about whether this is a local or remote ID, because I can see that leading to problems down the road. How about we rename this from "id" to "remoteId"?

wzieba replied:
I agree - for this reason I set type of `RemoteId` and not `Long` type. We already have `remoteId` field in this class of type `Long` - I'm not sure how differently we can handle this. I'd opt to leave `RemoteId` type - I believe this way we assert correct id more robustly. WDYT?

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt:
```diff
@@ -1,38 +1,31 @@
+    @Deprecated(replaceWith = ReplaceWith("id"), message = "Use id alias instead.")
     val remoteId: Long,
     val number: String,
-    val localSiteId: Int,
```
Are we sure it's safe to remove `localSiteId`? I'm just wondering if there will be situations where we need that to identify orders from sites other than the active one.

wzieba replied:
I think it's safe to remove it - it wasn't used anywhere and its more like an "inheritance" from database entity model - I think that in domain of Woo, an Order shouldn't know id of the site (especially its local id). If there will be a case to add this back, we can easily do this but for now I'd opt to remove it.

nbradbury commented:
@wzieba This all works fine, but there's one situation we're not covering and it's a bit of a mystery. If the user edits an order, returns to order detail, then leaves order detail before the request fails, the order will still show the edited information.

wzieba committed:
Update FluxC hash

wzieba committed:
Add localOrderId of LocalId type to Order. Deprecate other ids.

wzieba committed:
Align order update API to use local id

wzieba commented:
@nbradbury I've updated PR accordingly to our P2 discussion - it now uses `local order id` for identification. I've also updated hash of branch with "fake fail" so it's ready to perform retests with that.

About issue you mention:
> If the user edits an order, returns to order detail, then leaves order detail before the request fails, the order will still show the edited information.

That's true, good catch! I don't have a good solution for that from the top of my head. I was thinking about:

- blocking user to leave edit screen if some operations are in progress
- refresh (while showing cached) order on every enter in the order

but I dislike both.

I'll keep thinking about this case but I believe it's a non blocking issue - eventually after order refresh (swipe) user will see correct data. Do you agree or do you think we should address the issue right away?

nbradbury commented:
> I don't have a good solution for that from the top of my head. I was thinking about:
> * blocking user to leave edit screen if some operations are in progress
> * refresh (while showing cached) order on every enter in the order

I think we can rule out blocking the user, but refreshing the order seems good. But don't we already do that? At any rate, this doesn't need to hold up this PR.

wzieba committed:
Update FluxC hash

wzieba commented:
> don't we already do that?

I'm not sure if we do: there's `OrderDetailViewModel#checkIfFetchNeeded` method which causes refresh only if refunds do not match the server. I agree it'd be nice to consider silent refresh on every enter to details.

wzieba commented:
I've also updated FluxC hash to match `develop` branch @nbradbury.

nbradbury reviewed (approved):
Nice work on this! :shipit:
